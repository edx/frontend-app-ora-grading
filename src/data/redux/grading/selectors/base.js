import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import * as module from './base';

export const rootSelector = ({ grading }) => grading;
export const mkSimpleSelector = (key) => createSelector(
  [module.rootSelector],
  (grading) => grading[key],
);
export const rootKeys = StrictDict({
  selected: 'selected',
  activeIndex: 'activeIndex',
  current: 'current',
  gradeData: 'gradeData',
  gradingData: 'gradingData',
});

export const simpleSelectors = {
  selected: mkSimpleSelector(rootKeys.selected),
  activeIndex: mkSimpleSelector(rootKeys.activeIndex),
  current: mkSimpleSelector(rootKeys.current),
  gradeData: mkSimpleSelector(rootKeys.gradeData),
  gradingData: mkSimpleSelector(rootKeys.gradingData),
};

/**
 * returns the length of the list of selected submissions
 * @return {number} selected submission list length
 */
export const selectionLength = createSelector(
  [module.simpleSelectors.selected],
  (selected) => selected.length,
);

export default {
  ...simpleSelectors,
  selectionLength,
};
