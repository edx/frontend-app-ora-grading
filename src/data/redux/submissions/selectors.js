import _ from 'lodash';
import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import { lockStatuses } from 'data/services/lms/constants';

export const simpleSelectors = {
  allSubmissions: state => state.submissions.allSubmissions,
};

/**
 * Returns the submission list in default order for the table.
 */
export const listData = createSelector(
  [simpleSelectors.allSubmissions],
  (allSubmissions) => {
    const submissionIds = Object.keys(allSubmissions);
    const submissionList = submissionIds.map(id => {
      const { gradeStatus, lockStatus, ...rest } = allSubmissions[id];
      const gradingStatus = (lockStatus === lockStatuses.unlocked ? gradeStatus : lockStatus);
      return { gradingStatus, ...rest };
    });
    return _.sortBy(
      submissionList,
      ['submissionDate'],
    );
  },
);

export default StrictDict({
  ...simpleSelectors,
  listData,
});
