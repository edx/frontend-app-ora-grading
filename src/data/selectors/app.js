import { StrictDict } from 'utils';

export const simpleSelectors = {
  showReview: state => state.app.showReview,
  showRubric: state => state.app.showRubric,
  grading: state => state.app.grading,
  courseMetadata: state => state.app.courseMetadata,
  oraName: state => state.app.oraMetadata.name,
  oraPrompt: state => state.app.oraMetadata.prompt,
  oraTypes: state => state.app.oraMetadata.type,
};

export default StrictDict({
  ...simpleSelectors,
});
