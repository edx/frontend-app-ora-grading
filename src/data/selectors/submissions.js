import _ from 'lodash';
import { createSelector } from 'reselect';

import { StrictDict } from 'utils';

const simpleSelectors = {
  list: state => state.submissions.list,
  selected: state => state.submissions.selected,
  activeIndex: state => state.submissions.activeIndex,
  selectedResponse: state => state.submissions.current.response,
};

export const selectionIndex = createSelector(
  [simpleSelectors.list],
  (list, { learnerId }) => list.selected.indexOf(learnerId),
);

export const listData = createSelector(
  [simpleSelectors.list],
  (list) => _.sortBy(
    Object.values(list),
    ['learnerId'],
  ),
);

export const selectionLearnerId = (state, { index }) => (
  simpleSelectors.selected(state)[index].learnerId
);

export const currentSelection = createSelector(
  [simpleSelectors.selected, simpleSelectors.activeIndex],
  (list, activeIndex) => list[activeIndex],
);

export const selectedLearnerId = createSelector(
  [currentSelection],
  (selection) => selection.learnerId,
);

export const prevLearnerId = createSelector(
  [simpleSelectors.selected, simpleSelectors.activeIndex],
  (list, activeIndex) => {
    if (activeIndex > 0) {
      return list[activeIndex - 1].learnerId;
    }
    return null;
  },
);
export const nextLearnerId = createSelector(
  [simpleSelectors.selected, simpleSelectors.activeIndex],
  (list, activeIndex) => {
    if (activeIndex < list.length - 1) {
      return list[activeIndex + 1].learnerId;
    }
    return null;
  },
);

export default StrictDict({
  ...simpleSelectors,
  currentSelection,
  selectionIndex,
  listData,
  selectedLearnerId,
  nextLearnerId,
  prevLearnerId,
});
