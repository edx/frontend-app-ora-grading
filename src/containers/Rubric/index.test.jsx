import React from 'react';
import { shallow } from 'enzyme';

import selectors from 'data/selectors';
import { Rubric, mapStateToProps } from '.';

jest.mock('containers/CriterionContainer', () => 'CriterionContainer');
jest.mock('./RubricFeedback', () => 'RubricFeedback');

jest.mock('@edx/paragon', () => {
  const Card = () => 'Card';
  Card.Body = 'Card.Body';
  const Button = () => 'Button';
  return { Button, Card };
});

jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    app: {
      isGrading: jest.fn((...args) => ({ isGragrding: args })),
      rubric: {
        criteriaIndices: jest.fn((...args) => ({
          rubricCriteriaIndices: args,
        })),
      },
    },
  },
}));

describe('Rubric Container', () => {
  const props = {
    isGrading: true,
    criteriaIndices: [1, 2, 3, 4, 5],
  };
  let el;
  beforeEach(() => {
    el = shallow(<Rubric {...props} />);
  });
  describe('snapshot', () => {
    test('is grading', () => {
      expect(el).toMatchSnapshot();
    });
    test('is not grading', () => {
      el.setProps({
        isGrading: false,
      });
      expect(el).toMatchSnapshot();
    });
  });

  describe('component', () => {
    test('is grading (grading footer present)', () => {
      expect(el.find('.grading-rubric-footer').length).toEqual(1);
      const containers = el.find('CriterionContainer');
      expect(containers.length).toEqual(props.criteriaIndices.length);
      containers.forEach((container, i) => {
        expect(container.key()).toEqual(String(props.criteriaIndices[i]));
      });
    });

    test('is not grading (no grading footer)', () => {
      el.setProps({
        isGrading: false,
      });
      expect(el.find('.grading-rubric-footer').length).toEqual(0);
      const containers = el.find('CriterionContainer');
      expect(containers.length).toEqual(props.criteriaIndices.length);
      containers.forEach((container, i) => {
        expect(container.key()).toEqual(String(props.criteriaIndices[i]));
      });
    });
  });

  describe('mapStateToProps', () => {
    const testState = { abitaryState: 'some data' };
    let mapped;
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('selectors.app.isGrading', () => {
      expect(mapped.isGrading).toEqual(selectors.app.isGrading(testState));
    });

    test('selectors.app.rubric.criteriaIndices', () => {
      expect(mapped.criteriaIndices).toEqual(
        selectors.app.rubric.criteriaIndices(testState),
      );
    });
  });
});
