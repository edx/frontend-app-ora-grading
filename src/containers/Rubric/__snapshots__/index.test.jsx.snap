// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rubric Container snapshot is grading 1`] = `
<Card
  className="grading-rubric-card"
>
  <Card.Body
    className="grading-rubric-body"
  >
    <h3>
      <FormattedMessage
        defaultMessage="Rubric"
        description="Rubric interface label"
        id="ora-grading.Rubric.rubric"
      />
    </h3>
    <hr
      className="m-2.5"
    />
    <CriterionContainer
      isGrading={true}
      orderNum={1}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={2}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={3}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={4}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={5}
    />
    <hr />
    <RubricFeedback />
  </Card.Body>
  <div
    className="grading-rubric-footer"
  >
    <StatefulButton
      disabledStates={
        Array [
          "pending",
          "complete",
        ]
      }
      labels={
        Object {
          "complete": <FormattedMessage
            defaultMessage="Grade Submitted"
            description="Submit Grade button text after successful submission"
            id="ora-grading.Rubric.gradeSubmitted"
          />,
          "default": <FormattedMessage
            defaultMessage="Submit grade"
            description="Submit Grade button text"
            id="ora-grading.Rubric.submitGrade"
          />,
          "pending": <FormattedMessage
            defaultMessage="Submitting grade"
            description="Submit Grade button text while submitting"
            id="ora-grading.Rubric.submittingGrade"
          />,
        }
      }
      onClick={[MockFunction this.submitGradeHandler]}
      state="default"
    />
  </div>
</Card>
`;

exports[`Rubric Container snapshot is grading, lock is pending 1`] = `
<Card
  className="grading-rubric-card"
>
  <Card.Body
    className="grading-rubric-body"
  >
    <h3>
      <FormattedMessage
        defaultMessage="Rubric"
        description="Rubric interface label"
        id="ora-grading.Rubric.rubric"
      />
    </h3>
    <hr
      className="m-2.5"
    />
    <CriterionContainer
      isGrading={true}
      orderNum={1}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={2}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={3}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={4}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={5}
    />
    <hr />
    <RubricFeedback />
  </Card.Body>
  <div
    className="grading-rubric-footer"
  >
    <StatefulButton
      disabledStates={
        Array [
          "pending",
          "complete",
        ]
      }
      labels={
        Object {
          "complete": <FormattedMessage
            defaultMessage="Grade Submitted"
            description="Submit Grade button text after successful submission"
            id="ora-grading.Rubric.gradeSubmitted"
          />,
          "default": <FormattedMessage
            defaultMessage="Submit grade"
            description="Submit Grade button text"
            id="ora-grading.Rubric.submitGrade"
          />,
          "pending": <FormattedMessage
            defaultMessage="Submitting grade"
            description="Submit Grade button text while submitting"
            id="ora-grading.Rubric.submittingGrade"
          />,
        }
      }
      onClick={[MockFunction this.submitGradeHandler]}
      state="pending"
    />
  </div>
</Card>
`;

exports[`Rubric Container snapshot is grading, submit pending 1`] = `
<Card
  className="grading-rubric-card"
>
  <Card.Body
    className="grading-rubric-body"
  >
    <h3>
      <FormattedMessage
        defaultMessage="Rubric"
        description="Rubric interface label"
        id="ora-grading.Rubric.rubric"
      />
    </h3>
    <hr
      className="m-2.5"
    />
    <CriterionContainer
      isGrading={true}
      orderNum={1}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={2}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={3}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={4}
    />
    <CriterionContainer
      isGrading={true}
      orderNum={5}
    />
    <hr />
    <RubricFeedback />
  </Card.Body>
  <div
    className="grading-rubric-footer"
  >
    <StatefulButton
      disabledStates={
        Array [
          "pending",
          "complete",
        ]
      }
      labels={
        Object {
          "complete": <FormattedMessage
            defaultMessage="Grade Submitted"
            description="Submit Grade button text after successful submission"
            id="ora-grading.Rubric.gradeSubmitted"
          />,
          "default": <FormattedMessage
            defaultMessage="Submit grade"
            description="Submit Grade button text"
            id="ora-grading.Rubric.submitGrade"
          />,
          "pending": <FormattedMessage
            defaultMessage="Submitting grade"
            description="Submit Grade button text while submitting"
            id="ora-grading.Rubric.submittingGrade"
          />,
        }
      }
      onClick={[MockFunction this.submitGradeHandler]}
      state="pending"
    />
  </div>
</Card>
`;

exports[`Rubric Container snapshot is not grading 1`] = `
<Card
  className="grading-rubric-card"
>
  <Card.Body
    className="grading-rubric-body"
  >
    <h3>
      <FormattedMessage
        defaultMessage="Rubric"
        description="Rubric interface label"
        id="ora-grading.Rubric.rubric"
      />
    </h3>
    <hr
      className="m-2.5"
    />
    <CriterionContainer
      isGrading={false}
      orderNum={1}
    />
    <CriterionContainer
      isGrading={false}
      orderNum={2}
    />
    <CriterionContainer
      isGrading={false}
      orderNum={3}
    />
    <CriterionContainer
      isGrading={false}
      orderNum={4}
    />
    <CriterionContainer
      isGrading={false}
      orderNum={5}
    />
    <hr />
    <RubricFeedback />
  </Card.Body>
</Card>
`;

exports[`Rubric Container snapshot submit completed 1`] = `
<Card
  className="grading-rubric-card"
>
  <Card.Body
    className="grading-rubric-body"
  >
    <h3>
      <FormattedMessage
        defaultMessage="Rubric"
        description="Rubric interface label"
        id="ora-grading.Rubric.rubric"
      />
    </h3>
    <hr
      className="m-2.5"
    />
    <CriterionContainer
      isGrading={false}
      orderNum={1}
    />
    <CriterionContainer
      isGrading={false}
      orderNum={2}
    />
    <CriterionContainer
      isGrading={false}
      orderNum={3}
    />
    <CriterionContainer
      isGrading={false}
      orderNum={4}
    />
    <CriterionContainer
      isGrading={false}
      orderNum={5}
    />
    <hr />
    <RubricFeedback />
  </Card.Body>
  <div
    className="grading-rubric-footer"
  >
    <StatefulButton
      disabledStates={
        Array [
          "pending",
          "complete",
        ]
      }
      labels={
        Object {
          "complete": <FormattedMessage
            defaultMessage="Grade Submitted"
            description="Submit Grade button text after successful submission"
            id="ora-grading.Rubric.gradeSubmitted"
          />,
          "default": <FormattedMessage
            defaultMessage="Submit grade"
            description="Submit Grade button text"
            id="ora-grading.Rubric.submitGrade"
          />,
          "pending": <FormattedMessage
            defaultMessage="Submitting grade"
            description="Submit Grade button text while submitting"
            id="ora-grading.Rubric.submittingGrade"
          />,
        }
      }
      onClick={[MockFunction this.submitGradeHandler]}
      state="complete"
    />
  </div>
</Card>
`;
